src/components/PopupConfirmation.js

```diff
@@ -0,0 +1,36 @@
+import Popup from '../components/Popup.js';
+
+export default class PopupConfirmation extends Popup {
+  constructor(popup, handleDeletePicture) {
+    super(popup);
+
+    this._handleDeletePicture = handleDeletePicture;
+    this._popupForm = this._popup.querySelector('.popup__form');
+  }
+
+  open(cardId, element) {
+    this._cardId = cardId;
```

Comment:
Needs correction: `PopupWithConfirmation` is a general class. It can be used for any confirmation, not only card delete confirmation. This class should have a setter for `_handleFormSubmit`, not for card, so you can use a card in that handler in `index.js`.

```js
constructor(popupSelector) {
  super(popupSelector);
  this._form = this._modal.querySelector(".modal__form");
}

setAction(action) { 
  this._handleFormSubmit = action; 
} 

this._form.addEventListener("submit", (evt) => { 
  evt.preventDefault(); 
  this._handleFormSubmit(); 
})
```

Reply:
Hello! Thank you for your review! I have a couple of questions to ask. 
Firstly, I currently use a named function, which allows me to add listeners when the popup is opened and remove them after it is closed. If I implement these lines as suggested, I won't be able to remove the event listeners, resulting in a new listener being added each time the popup opens. What should I do in this case? Should I manually manage the editing of event listeners and add them in my index.js rather than incorporating them in the popup.open() method?

javascript
Copy code
this._form.addEventListener("submit", (evt) => { 
  evt.preventDefault(); 
  this._handleFormSubmit(); 
});
In my PopupWithConfirmation, I also modified the open() method so it passes the card ID and card element to the confirmation popup and the API. This modification has made my PopupWithConfirmation specific to card deletion confirmation. How could I pass this data to the confirmation and make PopupWithConfirmation versatile for various types of confirmations?

Reply:
Hello! To not affect your logic you can declare the submit listener as the class method.

See the following code:

```js
import Popup from '../components/Popup.js';

export default class PopupConfirmation extends Popup {
  constructor(popup, handleDeletePicture) {
    super(popup);
    this._popupForm = this._popup.querySelector('.popup__form');
  }

  // here we update the action
  // do that before opening the confirm popup for the specific card 
  setAction(action) { 
    this._action = action; 
  } 

  // here we declare submit handler
  _handleSubmitForm(event) {
    // prevent the default event behaviour
    event.preventDefault();
    // perform the action
    this._action();
    // note, that you should not close the popup here, you should do that only after the successful api response in index.js
  }

  setEventListeners() {
    this._popupForm.addEventListener('submit', this._handleSubmitForm);
    super.setEventListeners();
  }

  removeEventListeners() {
    this._popupForm.removeEventListener('submit', this._handleSubmitForm);
    super.removeEventListeners();
  }
}
```

Reply:
Thank you! However, I'm not entirely sure how to pass the cardId and card element to the handler while keeping my PopupConfirmation class general. Based on your suggestions, it seems I shouldn't use the action to pass the cardId and card element to index.js."
`
 open(cardId, element) {
// here I store my data
    this._cardId = cardId;
    this._element = element;
    super.open();
  }

  setAction(action) { 
    this._action = action; 
  }

_handleSubmitForm(event) {
    event.preventDefault();
// pass my data to index.js  this._action(this._cardId, this._element);
    this._action();
   index.js
  }
`

Reply:
You can pass the action right before opening the popup:

```
function handleDeleteButton(cardId, card) {
  popupDelete.setAction(() =>handleDeletePicture(cardId, card));
  popupDelete.open(t);
}
```